test/FunctionalTests/Infrastructure/AwaitableMemoryStream.cs

```diff
@@ -0,0 +1,110 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Grpc.AspNetCore.FunctionalTests.Infrastructure
+{
+    public class AwaitableMemoryStream : Stream
```

Comment:
Can we make it more clear what this class is for?
It took me a while to understand that the usage is to fill the stream with calling SendData() which then awakes the ReadAsync() operation - so basically this is a pipe. It sounds we could reuse some existing code for this purpose.

Btw, the logic in this class is nontrivial, so if left as is, shouldn't we have unit tests for this utility class?

Reply:
The implementation will be changed to `SyncPointMemoryStream` in  aspnet#16.